File: server/src/pricing.js
// Pricing engine — the single source of truth for how a line is priced.
// Pure ES module (no Node imports) so the same file is served to the tablet web app,
// and the Kotlin app implements the identical rules (see android-app/.../Pricing.kt).
//
// Rules
//  1. Base price = variant.price if a variant is chosen, else item.price.
//  2. For every option group:
//       - `included` = variant.included[group.id] if defined, else group.included (free picks).
//       - Picks are counted in selection order; picks beyond `included` cost group.extra_price each.
//       - Every selected option ALSO adds its own option.price (surcharge, e.g. smoked salmon +5.99).
//  3. Line total = (base + options) × qty.
//  4. Taxes are computed on the order subtotal: GST and QST both on the subtotal (Québec rule).

export function round2(n) {
  return Math.round((n + Number.EPSILON) * 100) / 100;
}

export function findItem(menu, itemId) {
  return menu.items.find(i => i.id === itemId) || null;
}

/**
 * @param {object} menu       full menu document
 * @param {object} line       { item_id, variant_id?, options: [{group_id, option_id}], qty, note? }
 * @param {string} lang       'fr' | 'en' — for the human-readable description
 * @returns {{ ok: boolean, error?: string, priced?: object }}
 */
export function priceLine(menu, line, lang = 'fr') {
  const item = findItem(menu, line.item_id);
  if (!item) return { ok: false, error: `Unknown item ${line.item_id}` };
  if (item.available === false) return { ok: false, error: `${name(item, lang)} is not available` };
  const cat = menu.categories.find(c => c.id === item.category_id);
  if (cat && cat.visible === false) return { ok: false, error: `${name(item, lang)} is not available` };

  const qty = Math.max(1, Math.min(50, parseInt(line.qty || 1, 10) || 1));
  let variant = null;
  if (item.variants && item.variants.length) {
    variant = item.variants.find(v => v.id === line.variant_id) || null;
    if (!variant) return { ok: false, error: `Please choose a size / formula for ${name(item, lang)}` };
  }
  let unit = variant ? variant.price : item.price;
  const chosen = [];   // [{group_id, option_id, name, price, extra}]
  const selections = Array.isArray(line.options) ? line.options : [];

  for (const g of item.option_groups || []) {
    const picks = selections.filter(s => s.group_id === g.id);
    const included = (variant && variant.included && g.id in variant.included) ? variant.included[g.id] : (g.included || 0);
    const type = g.type === 'single' ? 'single' : 'multi';
    if (type === 'single' && picks.length > 1) return { ok: false, error: `Only one choice allowed in ${name(g, lang)}` };
    const min = g.required ? Math.max(1, g.min || 0) : (g.min || 0);
    if (picks.length < min) return { ok: false, error: `${name(g, lang)}: please choose at least ${min}` };
    if (g.max != null && g.max > 0 && picks.length > g.max) return { ok: false, error: `${name(g, lang)}: maximum ${g.max}` };
    picks.forEach((p, idx) => {
      const o = (g.options || []).find(o => o.id === p.option_id);
      if (!o) throw Object.assign(new Error(`Unknown option ${p.option_id} in ${g.id}`), { user: true });
      if (o.available === false) throw Object.assign(new Error(`${name(o, lang)} is not available`), { user: true });
      const extra = idx >= included ? (g.extra_price || 0) : 0;
      const price = round2((o.price || 0) + extra);
      unit += price;
      chosen.push({ group_id: g.id, option_id: o.id, group_name: name(g, lang), name: name(o, lang), price, extra, surcharge: o.price || 0 });
    });
  }
  // Selections referring to groups the item doesn't have are ignored on purpose (stale app cache).
  unit = round2(unit);
  return {
    ok: true,
    priced: {
      item_id: item.id,
      category_id: item.category_id,
      name: name(item, lang),
      name_fr: item.name.fr, name_en: item.name.en || item.name.fr,
      variant_id: variant ? variant.id : null,
      variant_name: variant ? name(variant, lang) : null,
      options: chosen,
      qty,
      unit_price: unit,
      line_total: round2(unit * qty),
      note: (line.note || '').toString().slice(0, 200),
    },
  };
}

export function priceOrder(menu, lines, settings, lang = 'fr') {
  const priced = [];
  for (const l of lines) {
    let r;
    try { r = priceLine(menu, l, lang); }
    catch (e) { if (e.user) return { ok: false, error: e.message }; throw e; }
    if (!r.ok) return r;
    priced.push(r.priced);
  }
  if (!priced.length) return { ok: false, error: 'Empty order' };
  const subtotal = round2(priced.reduce((s, p) => s + p.line_total, 0));
  const gst = round2(subtotal * (Number(settings.tax_gst) || 0) / 100);
  const qst = round2(subtotal * (Number(settings.tax_qst) || 0) / 100);
  const total = round2(subtotal + gst + qst);
  return { ok: true, lines: priced, subtotal, tax_gst: gst, tax_qst: qst, total };
}

export function name(obj, lang) {
  if (!obj || !obj.name) return '';
  if (typeof obj.name === 'string') return obj.name;
  return (lang === 'en' && obj.name.en) ? obj.name.en : (obj.name.fr || obj.name.en || '');
}

export function formatPrice(n, lang = 'fr') {
  const v = (Math.round(n * 100) / 100).toFixed(2);
  return lang === 'en' ? `$${v}` : `${v.replace('.', ',')} $`;
}

File: android-app/app/src/main/java/ca/lartisan/menu/data/Pricing.kt
package ca.lartisan.menu.data

import kotlin.math.round

/**
 * Pricing rules — must stay identical to server/src/pricing.js (the server re-computes every order anyway,
 * this is only for what the customer sees before confirming).
 *
 *  1. Base price = variant.price if a variant is chosen, else item.price.
 *  2. For every option group: included = variant.included[group.id] ?: group.included.
 *     Picks are counted in selection order; picks beyond `included` cost group.extraPrice each.
 *     Every selected option also adds its own option.price (surcharge).
 *  3. Line total = (base + options) × qty.
 *  4. GST and QST are both computed on the subtotal.
 */
object Pricing {
    fun round2(v: Double): Double = round(v * 100.0) / 100.0

    data class PricedOption(val groupId: String, val optionId: String, val name: String, val price: Double, val extra: Double)
    data class PricedLine(
        val line: CartLine, val item: MenuItem, val variant: Variant?, val name: String, val variantName: String?,
        val options: List<PricedOption>, val unitPrice: Double, val lineTotal: Double,
    )
    data class PricedOrder(val lines: List<PricedLine>, val subtotal: Double, val gst: Double, val qst: Double, val total: Double)

    sealed class Result {
        data class Ok(val line: PricedLine) : Result()
        data class Error(val message: String) : Result()
    }

    fun includedFor(item: MenuItem, variant: Variant?, group: OptionGroup): Int =
        variant?.included?.get(group.id) ?: group.included

    fun priceLine(menu: Menu, line: CartLine, lang: String, s: Strings): Result {
        val item = menu.item(line.itemId) ?: return Result.Error(s.unavailable)
        if (!item.available) return Result.Error(s.unavailable)
        val variant = if (item.variants.isNotEmpty()) {
            item.variants.firstOrNull { it.id == line.variantId } ?: return Result.Error(s.chooseVariant)
        } else null
        var unit = variant?.price ?: item.price
        val chosen = mutableListOf<PricedOption>()
        for (g in item.optionGroups) {
            val picks = line.options.filter { it.groupId == g.id }
            val included = includedFor(item, variant, g)
            val min = if (g.required) maxOf(1, g.min) else g.min
            if (picks.size < min) return Result.Error("${g.name.get(lang)} : ${s.min(min)}")
            val max = g.max
            if (max != null && max > 0 && picks.size > max) return Result.Error("${g.name.get(lang)} : ${s.max(max)}")
            picks.forEachIndexed { idx, p ->
                val o = g.options.firstOrNull { it.id == p.optionId } ?: return Result.Error(s.unavailable)
                if (!o.available) return Result.Error("${o.name.get(lang)} — ${s.unavailable}")
                val extra = if (idx >= included) g.extraPrice else 0.0
                val price = round2(o.price + extra)
                unit += price
                chosen += PricedOption(g.id, o.id, o.name.get(lang), price, extra)
            }
        }
        unit = round2(unit)
        val qty = line.qty.coerceIn(1, 50)
        return Result.Ok(
            PricedLine(line, item, variant, item.name.get(lang), variant?.name?.get(lang), chosen, unit, round2(unit * qty))
        )
    }

    fun priceOrder(menu: Menu, lines: List<CartLine>, lang: String, s: Strings): PricedOrder {
        val priced = lines.mapNotNull { (priceLine(menu, it, lang, s) as? Result.Ok)?.line }
        val subtotal = round2(priced.sumOf { it.lineTotal })
        val gst = round2(subtotal * menu.settings.taxGst / 100.0)
        val qst = round2(subtotal * menu.settings.taxQst / 100.0)
        return PricedOrder(priced, subtotal, gst, qst, round2(subtotal + gst + qst))
    }

    fun money(v: Double, lang: String): String {
        val s = String.format(java.util.Locale.US, "%.2f", round2(v))
        return if (lang == "en") "$$s" else s.replace('.', ',') + " $"
    }
}
